# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
import pandas as pd
import os

import numpy as np


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. You can find the data in Ironhack's database (db: weather, table: austin_weather). 

In [40]:
austin = pd.read_csv(r'austin_weather.csv')

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [41]:
pd.set_option('display.max_columns', None)
austin

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0     2013-12-21         74        60        45            67           49   
1     2013-12-22         56        48        39            43           36   
2     2013-12-23         58        45        32            31           27   
3     2013-12-24         61        46        31            36           28   
4     2013-12-25         58        50        41            44           40   
...          ...        ...       ...       ...           ...          ...   
1314  2017-07-27        103        89        75            71           67   
1315  2017-07-28        105        91        76            71           64   
1316  2017-07-29        107        92        77            72           64   
1317  2017-07-30        106        93        79            70           68   
1318  2017-07-31         99        88        77            66           61   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
0              43                  93                 75                 57   
1              28                  93                 68                 43   
2              23                  76                 52                 27   
3              21                  89                 56                 22   
4              36                  86                 71                 56   
...           ...                 ...                ...                ...   
1314           61                  82                 54                 25   
1315           55                  87                 54                 20   
1316           55                  82                 51                 19   
1317           63                  69                 48                 27   
1318           54                  64                 43                 22   

     SeaLevelPressureHighInches SeaLevelPressureAvgInches  \
0                         29.86                     29.68   
1                         30.41                     30.13   
2                         30.56                     30.49   
3                         30.56                     30.45   
4                         30.41                     30.33   
...                         ...                       ...   
1314                      30.04                     29.97   
1315                      29.97                      29.9   
1316                      29.91                     29.86   
1317                      29.96                     29.91   
1318                      30.04                     29.97   

     SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
0                        29.59                  10                  7   
1                        29.87                  10                 10   
2                        30.41                  10                 10   
3                         30.3                  10                 10   
4                        30.27                  10                 10   
...                        ...                 ...                ...   
1314                     29.88                  10                 10   
1315                     29.81                  10                 10   
1316                     29.79                  10                 10   
1317                     29.87                  10                 10   
1318                     29.91                  10                 10   

     VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                     2          20          4          31   
1                     5          16          6          25   
2                    10           8          3          12   
3                     7          12          4          20   
4                     7          10          2          16   
...                 ...         ...        ...         ...   
1314                 10          12          5          21   
1315                

In [42]:
# we have 3 numeric columns
austin.describe().T

count       mean        std   min   25%   50%   75%    max
TempHighF  1319.0  80.862775  14.766523  32.0  72.0  83.0  92.0  107.0
TempAvgF   1319.0  70.642911  14.045904  29.0  62.0  73.0  83.0   93.0
TempLowF   1319.0  59.902957  14.190648  19.0  49.0  63.0  73.0   81.0

In [43]:
austin.dtypes

Date                          object
TempHighF                      int64
TempAvgF                       int64
TempLowF                       int64
DewPointHighF                 object
DewPointAvgF                  object
DewPointLowF                  object
HumidityHighPercent           object
HumidityAvgPercent            object
HumidityLowPercent            object
SeaLevelPressureHighInches    object
SeaLevelPressureAvgInches     object
SeaLevelPressureLowInches     object
VisibilityHighMiles           object
VisibilityAvgMiles            object
VisibilityLowMiles            object
WindHighMPH                   object
WindAvgMPH                    object
WindGustMPH                   object
PrecipitationSumInches        object
Events                        object
dtype: object

In [44]:
austin.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
PrecipitationSumInches        0
Events                        0
dtype: int64

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [45]:
austin.DewPointAvgF.unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [ ]:
# There is a negative symbol in the column. Which could mean NaN

In [46]:
austin[(df.DewPointAvgF == '-')]

Empty DataFrame
Columns: [Date, TempHighF, TempAvgF, TempLowF, DewPointHighF, DewPointAvgF, DewPointLowF, HumidityHighPercent, HumidityAvgPercent, HumidityLowPercent, SeaLevelPressureHighInches, SeaLevelPressureAvgInches, SeaLevelPressureLowInches, VisibilityHighMiles, VisibilityAvgMiles, VisibilityLowMiles, WindHighMPH, WindAvgMPH, WindGustMPH, PrecipitationSumInches, Events]
Index: []

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [18]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [47]:
austin[['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']] = austin[['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']].apply(pd.to_numeric, errors='coerce')

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [48]:
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [57]:
missing_values = austin[(austin.isnull().any(axis=1))]
missing_values.shape

(136, 21)

In [51]:
missing_values.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [63]:
austin.shape

(1319, 21)

#### Calculate the ratio of missing rows to total rows

In [61]:
136/1319*100

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [64]:
austin.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [66]:
austin.PrecipitationSumInches.count()   

1195

In [70]:
austin[(austin.PrecipitationSumInches.isna())].count()

Date                          124
TempHighF                     124
TempAvgF                      124
TempLowF                      124
DewPointHighF                 123
DewPointAvgF                  123
DewPointLowF                  123
HumidityHighPercent           124
HumidityAvgPercent            124
HumidityLowPercent            124
SeaLevelPressureHighInches    124
SeaLevelPressureAvgInches     124
SeaLevelPressureLowInches     124
VisibilityHighMiles           122
VisibilityAvgMiles            122
VisibilityLowMiles            122
WindHighMPH                   124
WindAvgMPH                    124
WindGustMPH                   124
PrecipitationSumInches          0
Events                        124
dtype: int64

In [71]:
124/1195*100

10.376569037656903

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [72]:
austin.drop(columns = 'PrecipitationSumInches', inplace = True)

In [75]:
austin

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

In [74]:
# Print `austin` to confirm the column is indeed removed

austin.PrecipitationSumInches

AttributeError: 'DataFrame' object has no attribute 'PrecipitationSumInches'

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [181]:
austin_fixed = austin.copy()

In [182]:
austin_fixed.interpolate(inplace=True)

In [183]:
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [184]:
austin.isna().sum()

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [185]:
austin_fixed.Events.value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Fog , Thunderstorm             1
Rain , Snow                    1
Name: Events, dtype: int64

In [186]:
austin_fixed.Events.nunique()

9

In [187]:
austin_fixed.Events.unique()

array(['Rain , Thunderstorm', ' ', 'Rain', 'Fog', 'Rain , Snow',
       'Fog , Rain', 'Thunderstorm', 'Fog , Rain , Thunderstorm',
       'Fog , Thunderstorm'], dtype=object)

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [188]:
# Your answer:  3


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [189]:
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

In [190]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
austin_fixed[event_list] = 0


# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [191]:
austin_fixed.loc[(austin_fixed.Events.str.contains('Rain')), 'Rain'] = 1
austin_fixed.loc[(austin_fixed.Events.str.contains('Snow')), 'Snow'] = 1
austin_fixed.loc[(austin_fixed.Events.str.contains('Fog')), 'Fog'] = 1
austin_fixed.loc[(austin_fixed.Events.str.contains('Thunderstorm')), 'Thunderstorm'] = 1

In [192]:
austin_fixed[(austin_fixed.Events == 'Rain')| (austin_fixed.Events == 'Snow')| (austin_fixed.Events == 'Fog')| (austin_fixed.Events == 'Thunderstorm')]

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
18    2014-01-08         53        47        40           51.0          45.0   
19    2014-01-09         70        62        53           60.0          55.0   
20    2014-01-10         72        65        57           64.0          61.0   
22    2014-01-12         67        57        46           58.0          47.0   
37    2014-01-27         60        47        34           43.0          23.0   
...          ...        ...       ...       ...            ...           ...   
1281  2017-06-24         85        79        72           76.0          71.0   
1282  2017-06-25         90        83        75           74.0          71.0   
1284  2017-06-27         91        81        71           74.0          72.0   
1293  2017-07-06        100        89        77           74.0          71.0   
1295  2017-07-08        100        88        75           71.0          67.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
18            30.0                 93.0                75.0   
19            50.0                 93.0                80.0   
20            54.0                 93.0                81.0   
22            33.0                 84.0                68.0   
37             5.0                 66.0                44.0   
...            ...                  ...                 ...   
1281          66.0                 85.0                70.0   
1282          67.0                 87.0                67.0   
1284          67.0                100.0                75.0   
1293          63.0                 88.0                60.0   
1295          62.0                 87.0                59.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
18                  57.0                       30.33   
19                  66.0                       30.18   
20                  68.0                       29.97   
22                  52.0                       30.12   
37                  21.0                       30.28   
...                  ...                         ...   
1281                55.0                       30.12   
1282                47.0                       30.19   
1284                50.0                       30.05   
1293                31.0                       30.15   
1295                31.0                       30.09   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
18                        30.20                      30.12   
19                        30.11                      30.00   
20                        29.87                      29.75   
22                        30.01                      29.90   
37                        30.09                      29.81   
...                         ...                        ...   
1281                      30.02                      29.89   
1282                      30.11                      30.05   
1284                      30.00                      29.90   
1293                      30.06                      29.98   
1295                      30.02                      29.94   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
18                   10.0                 2.0                 1.0   
19                   10.0                 4.0                 0.0   
20                   10.0                 8.0                 1.0   
22                   10.0                10.0                10.0   
37                   10.0                10.0                10.0   
...                   ...                 ...                 ...   
1281                 10.0                 8.0                 1.0   
1282                 10.0                10.0                 7.0   
1284                 10.0                 7.0                 2.0   
1293                 10.0                10.0                10.0   
1295                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [193]:
austin_fixed


Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH     

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [195]:
austin_fixed.drop(columns = 'Events', inplace = True)

austin_fixed.Events

KeyError: "['Events'] not found in axis"

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [196]:
from datetime import datetime 

In [197]:
austin_fixed.Date

0       2013-12-21
1       2013-12-22
2       2013-12-23
3       2013-12-24
4       2013-12-25
           ...    
1314    2017-07-27
1315    2017-07-28
1316    2017-07-29
1317    2017-07-30
1318    2017-07-31
Name: Date, Length: 1319, dtype: object

In [198]:
austin_fixed.Date=austin_fixed.Date.apply(lambda x : dt.strptime(x, '%Y-%M-%d').toordinal())

In [199]:
austin_fixed.Date.dtypes

dtype('int64')

#### Print `austin_fixed` to check your `Date` column.

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [200]:
y = austin_fixed.TempAvgF

X = austin_fixed[['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']]

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [201]:
from sklearn.model_selection import train_test_split 

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [202]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size= .8, test_size=.2)

In [203]:
X_train

Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
86    735250         68        41           31.0          26.0          20.0   
356   735245         71        59           58.0          57.0          56.0   
1316  736358        107        77           72.0          64.0          55.0   
1138  736330         82        51           59.0          52.0          45.0   
851   735983         81        60           67.0          63.0          58.0   
...      ...        ...       ...            ...           ...           ...   
1146  736338         68        43           37.0          31.0          28.0   
727   735616         58        34           35.0          31.0          27.0   
1068  735986         73        51           69.0          54.0          45.0   
36    735259         76        43           45.0          39.0          33.0   
279   735259         85        66           67.0          64.0          63.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
86                   53.0                36.0                18.0   
356                  90.0                78.0                66.0   
1316                 82.0                51.0                19.0   
1138                100.0                68.0                35.0   
851                  93.0                72.0                51.0   
...                   ...                 ...                 ...   
1146                 54.0                39.0                24.0   
727                  76.0                55.0                33.0   
1068                100.0                71.0                41.0   
36                   76.0                52.0                27.0   
279                  87.0                71.0                54.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
86                         30.06                      29.94   
356                        30.30                      30.23   
1316                       29.91                      29.86   
1138                       30.09                      29.98   
851                        30.06                      29.98   
...                          ...                        ...   
1146                       30.38                      30.28   
727                        30.45                      30.37   
1068                       30.21                      30.11   
36                         30.00                      29.84   
279                        30.15                      30.06   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
86                        29.78                 10.0                10.0   
356                       30.18                 10.0                10.0   
1316                      29.79                 10.0                10.0   
1138                      29.89                 10.0                 8.0   
851                       29.90                 10.0                 9.0   
...                         ...                  ...                 ...   
1146                      30.19                 10.0                10.0   
727                       30.28                 10.0                10.0   
1068                      29.96                 10.0                 9.0   
36                        29.69                 10.0                10.0   
279                       29.99                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
86                  10.0         16.0         6.0         26.0     0    0   
356                  9.0         12.0         5.0         17.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1138                 2.0         14.0         4.0         20.0     0    0   
851                  2.0         20.0         4.0         35.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1146     

In [204]:
X_test

Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
760   735983         57        45           51.0          39.0          31.0   
937   735978        101        78           75.0          71.0          64.0   
984   735994         94        74           74.0          72.0          70.0   
1099  735987         78        65           68.0          66.0          64.0   
342   735261         72        42           54.0          45.0          35.0   
...      ...        ...       ...            ...           ...           ...   
162   735234         88        71           71.0          69.0          63.0   
1258  736330         89        67           71.0          68.0          65.0   
671   735621         80        71           76.0          73.0          69.0   
284   735234         92        76           73.0          70.0          63.0   
913   735984         96        75           73.0          71.0          69.0   

      HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
760                  89.0                70.0                50.0   
937                  91.0                62.0                33.0   
984                 100.0                75.0                49.0   
1099                100.0                85.0                69.0   
342                  82.0                62.0                41.0   
...                   ...                 ...                 ...   
162                  93.0                69.0                45.0   
1258                 97.0                75.0                53.0   
671                 100.0                91.0                82.0   
284                  88.0                68.0                47.0   
913                  87.0                66.0                45.0   

      SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
760                        30.23                      30.10   
937                        30.05                      29.97   
984                        29.97                      29.93   
1099                       30.08                      30.02   
342                        30.38                      30.24   
...                          ...                        ...   
162                        29.97                      29.90   
1258                       30.04                      29.95   
671                        29.94                      29.89   
284                        29.87                      29.79   
913                        30.14                      30.08   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
760                       29.99                 10.0                10.0   
937                       29.86                 10.0                10.0   
984                       29.86                 10.0                10.0   
1099                      29.96                 10.0                 9.0   
342                       30.09                 10.0                10.0   
...                         ...                  ...                 ...   
162                       29.83                 10.0                 8.0   
1258                      29.86                 10.0                10.0   
671                       29.82                 10.0                 7.0   
284                       29.72                 10.0                10.0   
913                       29.98                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
760                  9.0         12.0         4.0         17.0     0    0   
937                  9.0         13.0         7.0         22.0     0    0   
984                  8.0          9.0         2.0         14.0     0    0   
1099                 5.0         17.0         7.0         28.0     0    0   
342                 10.0         14.0         4.0         24.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
162      

In [205]:
y_train

86      55
356     65
1316    92
1138    67
851     71
        ..
1146    56
727     46
1068    62
36      60
279     76
Name: TempAvgF, Length: 1055, dtype: int64

In [206]:
y_test

760     51
937     90
984     84
1099    72
342     57
        ..
162     80
1258    78
671     76
284     84
913     86
Name: TempAvgF, Length: 264, dtype: int64

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [207]:
austin_fixed.shape

(1319, 23)

In [213]:
perc = round(len(austin_fixed)/100*80)
perc

1055

In [214]:
ts_rows = austin_fixed.head(perc)
ts_rows

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     734889         74        60        45           67.0          49.0   
1     734890         56        48        39           43.0          36.0   
2     734891         58        45        32           31.0          27.0   
3     734892         61        46        31           36.0          28.0   
4     734893         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1050  735968         72        68        63           68.0          66.0   
1051  735969         76        68        59           67.0          60.0   
1052  735970         71        66        61           69.0          65.0   
1053  735971         69        66        62           66.0          64.0   
1054  735972         65        63        60           62.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1050          59.0                100.0                91.0   
1051          57.0                 93.0                85.0   
1052          61.0                100.0                94.0   
1053          62.0                100.0                92.0   
1054          58.0                100.0                94.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1050                81.0                       30.27   
1051                76.0                       30.19   
1052                87.0                       30.14   
1053                84.0                       30.21   
1054                87.0                       30.29   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1050                      30.22                      30.14   
1051                      30.13                      30.07   
1052                      30.09                      30.04   
1053                      30.16                      30.09   
1054                      30.23                      30.18   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1050                 10.0                 6.0                 1.0   
1051                 10.0                10.0                 5.0   
1052                 10.0                 5.0                 1.0   
1053                 10.0                 8.0                 2.0   
1054                 10.0                 6.0                 1.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0            20

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [223]:
perc_x_ts_train = round(len(ts_rows)/100*80) 

x_ts_train = ts_rows.head(perc_x_ts_train)
x_ts_train

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    734889         74        60        45           67.0          49.0   
1    734890         56        48        39           43.0          36.0   
2    734891         58        45        32           31.0          27.0   
3    734892         61        46        31           36.0          28.0   
4    734893         58        50        41           44.0          40.0   
..      ...        ...       ...       ...            ...           ...   
839  735971         80        68        55           49.0          43.0   
840  735972         77        68        58           58.0          53.0   
841  735973         79        72        65           66.0          62.0   
842  735974         89        80        70           71.0          68.0   
843  735975         73        68        62           70.0          61.0   

     DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0            43.0                 93.0                75.0   
1            28.0                 93.0                68.0   
2            23.0                 76.0                52.0   
3            21.0                 89.0                56.0   
4            36.0                 86.0                71.0   
..            ...                  ...                 ...   
839          38.0                 77.0                51.0   
840          44.0                 84.0                63.0   
841          59.0                 87.0                75.0   
842          66.0                 93.0                74.0   
843          58.0                 93.0                83.0   

     HumidityLowPercent  SeaLevelPressureHighInches  \
0                  57.0                       29.86   
1                  43.0                       30.41   
2                  27.0                       30.56   
3                  22.0                       30.56   
4                  56.0                       30.41   
..                  ...                         ...   
839                25.0                       30.19   
840                41.0                       30.16   
841                62.0                       30.04   
842                55.0                       29.98   
843                73.0                       30.13   

     SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                        29.68                      29.59   
1                        30.13                      29.87   
2                        30.49                      30.41   
3                        30.45                      30.30   
4                        30.33                      30.27   
..                         ...                        ...   
839                      30.11                      30.02   
840                      30.09                      29.99   
841                      29.96                      29.85   
842                      29.89                      29.83   
843                      30.03                      29.94   

     VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
0                   10.0                 7.0                 2.0         20.0   
1                   10.0                10.0                 5.0         16.0   
2                   10.0                10.0                10.0          8.0   
3                   10.0                10.0                 7.0         12.0   
4                   10.0                10.0                 7.0         10.0   
..                   ...                 ...                 ...          ...   
839                 10.0                10.0                10.0         12.0   
840                 10.0                10.0                 2.0         16.0   
841                 10.0                10.0                10.0         21.0   
842                 10.0                10.0                 6.0         15.0   
843                 10.0                 8.0                 1.0         15.

In [226]:
perc_X_ts_test = round(len(ts_rows)/100*20) 

X_ts_test = ts_rows.tail(perc_X_ts_test)
X_ts_test

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
844   735976         72        65        58           61.0          60.0   
845   735977         77        66        54           58.0          56.0   
846   735978         79        67        54           62.0          58.0   
847   735979         75        70        64           71.0          66.0   
848   735980         78        71        64           71.0          67.0   
...      ...        ...       ...       ...            ...           ...   
1050  735968         72        68        63           68.0          66.0   
1051  735969         76        68        59           67.0          60.0   
1052  735970         71        66        61           69.0          65.0   
1053  735971         69        66        62           66.0          64.0   
1054  735972         65        63        60           62.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
844           58.0                100.0                83.0   
845           53.0                100.0                75.0   
846           53.0                100.0                77.0   
847           61.0                100.0                92.0   
848           63.0                 97.0                90.0   
...            ...                  ...                 ...   
1050          59.0                100.0                91.0   
1051          57.0                 93.0                85.0   
1052          61.0                100.0                94.0   
1053          62.0                100.0                92.0   
1054          58.0                100.0                94.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
844                 66.0                       30.09   
845                 50.0                       30.07   
846                 54.0                       29.93   
847                 84.0                       29.89   
848                 82.0                       29.99   
...                  ...                         ...   
1050                81.0                       30.27   
1051                76.0                       30.19   
1052                87.0                       30.14   
1053                84.0                       30.21   
1054                87.0                       30.29   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
844                       30.04                      29.95   
845                       30.00                      29.91   
846                       29.87                      29.78   
847                       29.84                      29.78   
848                       29.93                      29.85   
...                         ...                        ...   
1050                      30.22                      30.14   
1051                      30.13                      30.07   
1052                      30.09                      30.04   
1053                      30.16                      30.09   
1054                      30.23                      30.18   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
844                  10.0                 8.0                 2.0   
845                  10.0                 6.0                 0.0   
846                  10.0                 8.0                 4.0   
847                  10.0                 7.0                 1.0   
848                  10.0                 5.0                 0.0   
...                   ...                 ...                 ...   
1050                 10.0                 6.0                 1.0   
1051                 10.0                10.0                 5.0   
1052                 10.0                 5.0                 1.0   
1053                 10.0                 8.0                 2.0   
1054                 10.0                 6.0                 1.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
844          10

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [228]:
perc_y_ts_train = round(len(ts_rows.TempAvgF)/100*80)

y_ts_train = ts_rows.TempAvgF.head(perc_y_ts_train)
y_ts_train

0      60
1      48
2      45
3      46
4      50
       ..
839    68
840    68
841    72
842    80
843    68
Name: TempAvgF, Length: 844, dtype: int64

In [229]:
perc_y_ts_test = round(len(ts_rows.TempAvgF)/100*20)

y_ts_test = ts_rows.TempAvgF.tail(perc_y_ts_train)
y_ts_test

211     84
212     87
213     86
214     86
215     84
        ..
1050    68
1051    68
1052    66
1053    66
1054    63
Name: TempAvgF, Length: 844, dtype: int64